[Refactor](nereids) Abstract interface of statistics framework for new optimizer reuse

Kikyou1997 opened the pull request:
# Proposed changes

As the statistics framework could not be reused by new optmizer before, so I abstract some interface to make it reusable.

1. Make Slot extends the Id
2. Add new interface:ExprStats,PlanStats
3. Move definition of PlanNode.NodeType to statistics sub-directory

## Checklist(Required)

1. Does it affect the original behavior: (No)
5. Has unit tests been added: (No Need)
6. Has document been added or modified: (No)
7. Does it need to update dependencies: (No)
8. Are there any changes that cannot be rolled back: (No)

Kikyou1997 committed:
Refactor some code

Kikyou1997 committed:
add license

morrySnow commented on fe/fe-core/src/main/java/org/apache/doris/nereids/trees/expressions/ExprId.java:
```diff
@@ -17,13 +17,15 @@
  */
-public class ExprId {
+public class ExprId extends Id<ExprId> {
     private final long id;
     private final UUID jvmId;
```
remove these lines instead with parent class's attribute: id

Kikyou1997 replied:
done

Kikyou1997 replied:
And removed the jvmId by the way

924060929 commented on fe/fe-core/src/main/java/org/apache/doris/nereids/trees/plans/PlaceHolderPlan.java:
```diff
@@ -81,4 +85,34 @@ public PlaceHolderPlan withChildren(List<Plan> children) {
+    }
+
+    @Override
+    public StatisticalType getNodeType() {
```
```suggestion
    public StatisticalType getStatisticalType() {
```

Kikyou1997 replied:
done

morrySnow commented on fe/fe-core/src/main/java/org/apache/doris/planner/PlanNode.java:
```diff
@@ -353,6 +327,11 @@ public List<Expr> getConjuncts() {
 
+    @Override
+    public List<? extends PlanStats> getChildrenStats() {
+        return children;
```
why return children here? i think this interface should be used to return children's StatsDeriveResult

Kikyou1997 replied:
It chould get StatsDeriveResult by `PlanStats::getStatsDeriveResult`.

Kikyou1997 replied:
done

morrySnow commented on fe/fe-core/src/main/java/org/apache/doris/statistics/StatsRecursiveDerive.java:
```diff
@@ -38,11 +36,11 @@ private static class Inner {
      * This parameter is an input and output parameter,
      * which will store the derivation result of statistical information in the corresponding node
      */
-    public void statsRecursiveDerive(PlanNode node) throws UserException {
```
statsRecursiveDerive's parameter should not change to PlanStats, since Nereids will never use this to do stats Derive

Kikyou1997 committed:
1. remove jvmId in ExprId
2. rename PlanStats::getNodeType to PlanStats::getStatisticalType
3. throw Unsupported Method RuntimeException in PlaceHolderPlan.java
4. recover StatsRecursiveDerive.java

morrySnow commented on fe/fe-core/src/main/java/org/apache/doris/statistics/PlanStats.java:
```diff
@@ -0,0 +1,39 @@
+ */
+public interface PlanStats {
+
+    List<? extends PlanStats> getChildrenStats();
```
this should return children's StatsDeriveResult

Kikyou1997 committed:
List<? extends PlanStats> getChildrenStats(); -> List<StatsDeriveResult> getChildrenStats();

morrySnow reviewed (approved):
LGTM

morningman reviewed (approved):
LGTM
